src/nni_manager/training_service/kubeflow/azureStorageClientUtils.ts

```diff
@@ -0,0 +1,183 @@
+                deferred.reject(error);
+            }else{          
+                deferred.resolve();
+            }
+        })
+        return deferred.promise;
+    }
+    
+    /**
+     * download a file from azure storage
+     * @param fileServerClient 
+     * @param azureDirectory 
+     * @param azureFileName 
+     * @param azureShare 
+     * @param localFilePath 
+     */
+    async function downloadFile(fileServerClient: any, azureDirectory: any, azureFileName: any, azureShare: any, localFilePath: any){
+        const deferred: Deferred<void> = new Deferred<void>();
+        await fileServerClient.getFileToStream(azureShare, azureDirectory, azureFileName, fs.createWriteStream(localFilePath), function(error: any, result: any, response: any) {
+            if(error){
+                getLogger().error(`Download file failed:, ${error}`);
+                deferred.reject(error);
+            }else{
+                deferred.resolve();    
+            }
+        })
+        return deferred.promise;
+    }
+
+    /**
+     * Upload a directory to azure file storage
+     * @param fileServerClient 
+     */
+    export async function uploadDirectory(fileServerClient: any, azureDirectory: any, azureShare: any, localDirectory: any): Promise<void>{
+        const deferred: Deferred<void> = new Deferred<void>();
+        const fileNameArray: string[] = fs.readdirSync(localDirectory);
+        await createDirectoryRecursive(fileServerClient, azureDirectory, azureShare);
+        for(let fileName of fileNameArray){
+            const fullFilePath: string = path.join(localDirectory, fileName);
+            try {
+                if (fs.lstatSync(fullFilePath).isFile()) {
+                    await uploadFileToAzure(fileServerClient, azureDirectory, fileName, azureShare, fullFilePath);
+                } else {
+                    // If filePath is a directory, recuisively copy it to azure
+                    await uploadDirectory(fileServerClient, azureDirectory + '/' + fileName, azureShare, fullFilePath);
+                }
+            } catch(error) {
+                deferred.reject(error);
+                return deferred.promise;
+            }
+        }
+        // All files/directories are copied successfully, resolve
+        deferred.resolve();
+        return deferred.promise;
+    }
+    
+    /**
+     * downlod a directory from azure
+     * @param fileServerClient 
+     * @param azureDirectory 
+     * @param azureShare 
+     * @param localDirectory 
+     */
+    export async function downloadDirectory(fileServerClient: any, azureDirectory:any, azureShare: any, localDirectory: any): Promise<void>{
+        const deferred: Deferred<void> = new Deferred<void>();
+        mkDirP(localDirectory);
+        fileServerClient.listFilesAndDirectoriesSegmented(azureShare, azureDirectory, 'null', function(error: any, result: any, response: any) {
+            for(var fileName of result['entries']['files']){
```

Comment:
suggest to check if result has 'entries' key and 'file's key first, otherwise will throw exception